File: haskell/strain/src/Strain.hs
module Strain (keep, discard) where

keep :: (a -> Bool) -> [a] -> [a]
keep _ [] = []
keep p (x:xs)
  | p x = x : keep p xs
  | otherwise = keep p xs

discard :: (a -> Bool) -> [a] -> [a]
discard p = keep (not . p)

File: clojure/strain/src/strain.clj
(ns strain
  (:use [clojure.core]))

(defn retain [f seq]
  (loop [seq seq
         acc []]
    (cond (empty? seq) acc
          (f (first seq)) (recur (rest seq) (conj acc (first seq)))
          :else (recur (rest seq) acc))))

(defn discard [f seq]
  (loop [seq seq
         acc []]
    (cond (empty? seq) acc
          (f (first seq)) (recur (rest seq) acc)
          :else (recur (rest seq) (conj acc (first seq))))))